server/server_test.go

```diff
@@ -770,6 +774,138 @@ func TestLogSourceIPs(t *testing.T) {
 	require.Equal(t, fake.sourceIPs, "127.0.0.1")
 }
 
+func TestHTTPGRPCInstrumentationTracing(t *testing.T) {
+	tracer, closer := jaeger.NewTracer(
+		"test",
+		jaeger.NewConstSampler(true),
+		jaeger.NewInMemoryReporter(),
+	)
+	opentracing.SetGlobalTracer(tracer)
+	defer closer.Close()
+
+	var cfg Config
+	cfg.HTTPListenPort = 9099
+	cfg.GRPCListenAddress = "localhost"
+	cfg.GRPCListenPort = 1234
+	cfg.GPRCServerMaxRecvMsgSize = 4 * 1024 * 1024
+	cfg.GRPCServerMaxSendMsgSize = 4 * 1024 * 1024
+	cfg.Router = middleware.InitHTTPGRPCMiddleware(mux.NewRouter())
+	cfg.MetricsNamespace = "testing_httpgrpc_tracing"
+
+	server, err := New(cfg)
+	require.NoError(t, err)
+
+	type tracingSpanJSON struct {
+		OperationName string
+		Tags          map[string]interface{}
+	}
+
+	// writeSpanToBodyHandleFunc provides a way to test tracing by dumping the data
+	// from the requests' currently-active span data into a structured JSON body.
+	//
+	// HTTPGRPCTracer's parent Span tagging cannot be tested with current tracer implementations;
+	// tagging the parentSpan depends on methods only available on the Jaeger tracer, which limits
+	// introspection in tests as jaeger.Tracer can only access the current span, not the parent.
+	//
+	// MockTracer from opentracing/opentracing-go provides visibility into all completed spans for
+	// testing, but does not implement the jaeger operations needed to tag the parent span
+	writeSpanToBodyHandleFunc := func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Can't we inspect the created spans on `jaeger.InMemoryReporter` with `GetSpans()`?

Reply:
after a ton of digging, I found that you can inject an `Observer` which uses `SpanObserver`s to collect all this - latest approach is up now

Reply:
Does inspecting the `InMemoryReporter` we instantiate when creating the `Tracer` not work? Seems like it would be simpler than creating our own `Observer` type.

Reply:
No, the Reporter interface just produces `opentracing.Span`s which don't expose getter methods for anything about the span except `BaggageItems` - so you cannot ask an `opentracing.Span` implementation for the operation name or the tags.

So you have to mock and inject something, the observer is most straightforward it seems. You could mock the spans themselves, but that seems sketchy when you are trying to test their behavior in the first place.

Reply:
Right, makes sense, thanks for clarifying.